src/Microsoft.IdentityModel.Tokens/EventBasedLRUCache.cs

```diff
@@ -0,0 +1,332 @@
+            {
+                if (_eventQueue.TryTake(out var action))
+                {
+                    action.Invoke();
+                    if (ProcessingDelay != 0)
+                        Thread.Sleep(ProcessingDelay);
+                }
+            }
+        }
+
+        public bool Contains(TKey key)
+        {
+            if (key == null)
+                throw LogHelper.LogArgumentNullException(nameof(key));
+
+            return _map.ContainsKey(key);
+        }
+
+        /// <summary>
+        /// FOR TESTING PURPOSES ONLY.
+        /// </summary>
+        internal void WaitForProcessing()
+        {
+            while (_eventQueue.Count != 0)
+                continue;
+
+            return;
+        }
+
+        internal int RemoveExpiredValues()
+        {
+            int numItemsRemoved = 0;
+            var node = _doubleLinkedList.First;
+            while (node != null)
+            {
+                var nextNode = node.Next;
+                if (node.Value.ExpirationTime < DateTime.UtcNow)
+                {
+                    _doubleLinkedList.Remove(node);
+                    _map.TryRemove(node.Value.Key, out _);
+                    numItemsRemoved++;
+                }
+
+                node = nextNode;
+            }
+
+            return numItemsRemoved;
+        }
+
+        async Task RemoveExpiredValuesPeriodically(TimeSpan interval)
+        {
+            while (true)
+            {
+                _eventQueue.Add(() => RemoveExpiredValues());
+                await Task.Delay(interval).ConfigureAwait(false);
+            }
+        }
+
+        public void SetValue(TKey key, TValue value)
+        {
+            SetValue(key, value, DateTime.MaxValue);
+        }
+
+        public bool SetValue(TKey key, TValue value, DateTime expirationTime)
+        {
+            if (key == null)
+                throw LogHelper.LogArgumentNullException(nameof(key));
+
+            if (value == null)
+                throw LogHelper.LogArgumentNullException(nameof(value));
+
+            // item already expired
+            if (expirationTime < DateTime.UtcNow)
+                return false;
+
+            // just need to update value and move it to the top
+            if (_map.TryGetValue(key, out var cacheItem))
+            {
+                cacheItem.Value = value;
+                cacheItem.ExpirationTime = expirationTime;
+                _eventQueue.Add(() =>
+                {
+                    _doubleLinkedList.Remove(cacheItem);
+                    _doubleLinkedList.AddFirst(cacheItem);
+                });
+            }
+            else
+            {
+                // if cache is full, then remove the least recently used node
+                if (_map.Count >= _capacity)
+                {
+                    _eventQueue.Add(() =>
+                    {
+                        var lru = _doubleLinkedList.Last;
+                        _map.TryRemove(lru.Value.Key, out _);
```

Comment:
>_map.TryRemove(lru.Value.Key, out _); [](start = 24, length = 37)

Suggestion: Perhaps more than 1 should be removed? Perhaps 5% or 10% of the capacity?